File: 2015/Day15/main.py
import re
from itertools import product

def parse_input(filename):
    ingredients = {}
    with open(filename, 'r') as file:
        for line in file:
            name, properties = line.strip().split(': ')
            ingredients[name] = list(map(int, re.findall(r'-?\d+', properties)))
    return ingredients

def calculate_score(ingredients, amounts):
    totals = [0] * 5
    for ingredient, amount in zip(ingredients.values(), amounts):
        for i in range(5):
            totals[i] += ingredient[i] * amount
    
    score = 1
    for total in totals[:-1]:
        score *= max(total, 0)
    
    return score

def calculate_score2(ingredients, amounts):
    totals = [0] * 5
    for ingredient, amount in zip(ingredients.values(), amounts):
        for i in range(5):
            totals[i] += ingredient[i] * amount
    
    if totals[4] != 500:
        return 0
    
    score = 1
    for total in totals[:-1]:
        score *= max(total, 0)
    
    return score

def find_best_score(ingredients):
    best_score = 0
    for amounts in product(range(101), repeat=len(ingredients)):
        if sum(amounts) == 100:
            score = calculate_score(ingredients, amounts)
            best_score = max(best_score, score)
    return best_score

def find_best_cookie(ingredients):
    best_score = 0
    for amounts in product(range(101), repeat=len(ingredients)):
        if sum(amounts) == 100:
            score = calculate_score2(ingredients, amounts)
            best_score = max(best_score, score)
    return best_score

ingredients = parse_input('input.txt')
print(f"Part 1: {find_best_score(ingredients)}")
print(f"Part 2: {find_best_cookie(ingredients)}")

File: 2015/Day15/main.dart
import 'dart:io';
import 'dart:math';

typedef Ingredients = Map<String, List<int>>;
typedef Amounts = List<int>;

void main() async {
  final ingredients = await parseInput('input.txt');
  print('Part 1: ${findBestScore(ingredients)}');
  print('Part 2: ${findBestCookie(ingredients)}');
}

int calculateScore(Ingredients ingredients, Amounts amounts) {
  final totals = List<int>.filled(5, 0);

  for (var i = 0; i < ingredients.values.length; i++) {
    final ingredient = ingredients.values.elementAt(i);
    final amount = amounts[i];
    for (var j = 0; j < 5; j++) {
      totals[j] += ingredient[j] * amount;
    }
  }

  var score = 1;
  for (var i = 0; i < 4; i++) {
    score *= max(0, totals[i]);
  }

  return score;
}

int calculateScore2(Ingredients ingredients, Amounts amounts) {
  final totals = List<int>.filled(5, 0);

  for (var i = 0; i < ingredients.values.length; i++) {
    final ingredient = ingredients.values.elementAt(i);
    final amount = amounts[i];
    for (var j = 0; j < 5; j++) {
      totals[j] += ingredient[j] * amount;
    }
  }

  if (totals[4] != 500) {
    return 0;
  }

  var score = 1;
  for (var i = 0; i < 4; i++) {
    score *= max(0, totals[i]);
  }

  return score;
}

int findBestCookie(Ingredients ingredients) {
  var bestScore = 0;

  for (var amounts in generateCombinations(ingredients.length, 100)) {
    final score = calculateScore2(ingredients, amounts);
    bestScore = max(bestScore, score);
  }

  return bestScore;
}

int findBestScore(Ingredients ingredients) {
  var bestScore = 0;

  for (var amounts in generateCombinations(ingredients.length, 100)) {
    final score = calculateScore(ingredients, amounts);
    bestScore = max(bestScore, score);
  }

  return bestScore;
}

Iterable<Amounts> generateCombinations(int length, int total) sync* {
  if (length == 1) {
    yield [total];
  } else {
    for (var i = 0; i <= total; i++) {
      for (var combination in generateCombinations(length - 1, total - i)) {
        yield [i, ...combination];
      }
    }
  }
}

Future<Ingredients> parseInput(String filename) async {
  Ingredients ingredients = {};
  final lines = await File(filename).readAsLines();

  final regex = RegExp(r'-?\d+');

  for (var line in lines) {
    final parts = line.split(': ');
    final name = parts[0];
    final properties =
        regex.allMatches(parts[1]).map((m) => int.parse(m.group(0)!)).toList();
    ingredients[name] = properties;
  }

  return ingredients;
}
